### PR TITLE
Bugfix for environment paths with backslashes

### PR DESCRIPTION
**Pull request by theseanything**
This PR make sure Environment objects being given a path with forward slashes. Fixes a bug whereby launching nested environments in a Windows environment would fail due to environment paths containing backslashes.

### CODE REVIEW

**Comment by theseanything** [comment]
Closed - as introduced a delete-env bug

File: sceptre/environment.py [hunk 1]
```diff
@@ -369,19 +372,18 @@ def _get_available_stacks(self):
...
-            basename for basename in base_config_file_names
-            if basename != "config"
+            os.path.splitext(os.path.basename(config_file))[0]
+            for config_file in config_files
         ]
-        available_stacks = [
-            "/".join([self.path, basename])
-            for basename in stack_basenames
+
+        if "config" in stack_basenames:
```
Comment (jamesroutley) [inline sceptre/environment.py]: `stack_basenames` doesn't word as a variable name if it includes a `config` item. Can we move this if statement to a filter in the list comp on 375? Or change the name

File: sceptre/environment.py [hunk 2]
```diff
@@ -74,20 +72,25 @@ def __repr__(self):
...
         Raises an InvalidEnvironmentPathError if the path has a leading or
         trailing slash.
 
         :param path: A directory path
         :type path: str
         :raises: sceptre.exceptions.InvalidEnvironmentPathError
+        :returns: A normalised path with forward slashes.
+        :returns: string
         """
+        path = path.replace('\\', '/')
```
Comment (jamesroutley) [inline sceptre/environment.py]: Could we add a comment as to why we're doing this - I don't think the code is self-explanatory

File: sceptre/environment.py [hunk 3]
```diff
@@ -74,20 +72,25 @@ def __repr__(self):
...
         """
+        Converts backslashes to forward slashes.
         Raises an InvalidEnvironmentPathError if the path has a leading or
         trailing slash.
 
         :param path: A directory path
         :type path: str
         :raises: sceptre.exceptions.InvalidEnvironmentPathError
+        :returns: A normalised path with forward slashes.
+        :returns: string
```
Comment (jamesroutley) [inline sceptre/environment.py]: I think this should be `rtype`

File: tests/test_environment.py [hunk 4]
```diff
@@ -74,21 +75,42 @@ def test_repr(self):
...
-            self.environment._check_env_path_valid(
+            self.environment._validate_path(
                 "/this/environment/path/is/invalid"
             )
 
-    def test_check_env_path_valid_with_invalid_path_trailing_slash(self):
+    def test_validate_path_with_leading_backslash(self):
+        with pytest.raises(InvalidEnvironmentPathError):
+            self.environment._validate_path(
+                '\\this\environment\path\is\invalid'
```
Comment (jamesroutley) [inline tests/test_environment.py]: Can we double quote strings

File: tests/test_environment.py [hunk 5]
```diff
@@ -74,21 +75,42 @@ def test_repr(self):
...
         with pytest.raises(InvalidEnvironmentPathError):
-            self.environment._check_env_path_valid(
+            self.environment._validate_path(
                 "this/environment/path/is/invalid/"
             )
 
+    def test_validate_path_with_trailing_backslash(self):
+        with pytest.raises(InvalidEnvironmentPathError):
+            self.environment._validate_path(
+                'this\environment\path\is\invalid\\'
```
Comment (jamesroutley) [inline tests/test_environment.py]: Why do we need to escape the trailing slash (if I understand what's happening here), but not the others?

Comment (jamesroutley) [inline tests/test_environment.py]: Oh it's so as to not escape the closing quote - I get it.

File: sceptre/environment.py (hunk 2 above)
Comment (jamesroutley) [inline sceptre/environment.py]: Do we need to worry about environment paths that contain single letter directories which happen to form special chars?

e.g.
```
In [12]: "\a\n\c".replace("\\", "/")
Out[12]: '\x07\n/c'
```

Comment (theseanything) [inline sceptre/environment.py]: Added comment in the doctring.

Comment (theseanything) [inline sceptre/environment.py]: So above you are using a string literal and should be aware that you haven't escaped your backslashes. When paths for Windows are read by python - they will have escaped backslashes e.g. "\\a\\n\\c".

Comment (theseanything) [inline sceptre/environment.py]: I mean e.g. "\\\\a\\\\n\\\\c" , github escaped my slashes ahah

**Pull request merged**